File: lib/pull_preview/list.rb
require 'terminal-table'

module PullPreview
  class List
    def self.run(opts)
      raise Error, "Invalid org/repo given" if opts.arguments.none?
      org, repo = opts.arguments.first.split("/", 2)

      table = Terminal::Table.new(headings: [
        "Name",
        "IP",
        "Size",
        "Region",
        "AZ",
        "Created on",
        "Tags",
      ])
      tags_to_find = {
        "stack" => STACK_NAME,
      }
      tags_to_find.merge!("repo_name" => repo) if repo
      tags_to_find.merge!("org_name" => org) if org

      PullPreview.provider.list_instances(tags: tags_to_find) do |instance|
        table << [
          instance.name,
          instance.public_ip,
          instance.size,
          instance.region,
          instance.zone,
          instance.created_at.iso8601,
          instance.tags.map{|tag| [tag.key, tag.value].join(":")}.join(","),
        ]
      end
      puts table
    end
  end
end

File: internal/pullpreview/list.go
package pullpreview

import (
	"errors"
	"fmt"
)

func RunList(opts ListOptions, provider Provider, logger *Logger) error {
	if opts.Org == "" && opts.Repo == "" {
		return errors.New("invalid org/repo given")
	}
	tags := map[string]string{
		"stack": StackName,
	}
	if opts.Repo != "" {
		tags["repo_name"] = opts.Repo
	}
	if opts.Org != "" {
		tags["org_name"] = opts.Org
	}
	instances, err := provider.ListInstances(tags)
	if err != nil {
		return err
	}
	fmt.Printf("Name\tIP\tSize\tRegion\tAZ\tCreated On\tTags\n")
	for _, inst := range instances {
		fmt.Printf("%s\t%s\t%s\t%s\t%s\t%s\t%v\n", inst.Name, inst.PublicIP, inst.Size, inst.Region, inst.Zone, inst.CreatedAt.Format("2006-01-02T15:04:05Z"), inst.Tags)
	}
	return nil
}
